File: app/src/main/java/com/yuk/miuiHomeR/utils/PrefsUtils.java
package com.yuk.miuiHomeR.utils;

import android.content.Context;
import android.content.SharedPreferences;

import java.io.File;
import java.lang.reflect.Field;

public class PrefsUtils {

    public static String mPrefsName = "MiuiHome_Config";
    public static String mPrefsPathCurrent = null;
    public static String mPrefsFileCurrent = null;
    public static SharedPreferences mSharedPreferences = null;
    public static String mPrefsPath = "/data/user_de/0/" + Helpers.mAppModulePkg + "/shared_prefs";
    public static String mPrefsFile = mPrefsPath + "/" + mPrefsName + ".xml";

    public static SharedPreferences getSharedPrefs(Context context, boolean protectedStorage, boolean multiProcess) {
        if (protectedStorage) context = Helpers.getProtectedContext(context);
        try {
            return context.getSharedPreferences(mPrefsName, multiProcess ? Context.MODE_MULTI_PROCESS | Context.MODE_WORLD_READABLE : Context.MODE_WORLD_READABLE);
        } catch (Throwable t) {
            return context.getSharedPreferences(mPrefsName, multiProcess ? Context.MODE_MULTI_PROCESS | Context.MODE_PRIVATE : Context.MODE_PRIVATE);
        }
    }

    public static SharedPreferences getSharedPrefs(Context context, boolean protectedStorage) {
        return getSharedPrefs(context, protectedStorage, false);
    }

    public static String getSharedPrefsPath() {
        if (mPrefsPathCurrent == null) try {
            Field mFile = mSharedPreferences.getClass().getDeclaredField("mFile");
            mFile.setAccessible(true);
            mPrefsPathCurrent = ((File)mFile.get(mSharedPreferences)).getParentFile().getAbsolutePath();
            return mPrefsPathCurrent;
        } catch (Throwable t) {
            System.out.print("Test" + t);
            return mPrefsPath;
        } else return mPrefsPathCurrent;
    }

    public static String getSharedPrefsFile() {
        if (mPrefsFileCurrent == null) try {
            Field fFile = mSharedPreferences.getClass().getDeclaredField("mFile");
            fFile.setAccessible(true);
            mPrefsFileCurrent = ((File)fFile.get(mSharedPreferences)).getAbsolutePath();
            System.out.println("Test: mPrefsFileCurrent");
            return mPrefsFileCurrent;
        } catch (Throwable t) {
            System.out.println("Test: mPrefsFile" + t);
            return mPrefsFile;
        } else
            System.out.println("Test: mPrefsFileCurrent2");
        return mPrefsFileCurrent;
    }
}

File: app/src/main/java/com/yuk/miuiHomeR/utils/PrefsUtils.kt
package com.yuk.miuiHomeR.utils

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import com.yuk.miuiHomeR.utils.Helpers.getProtectedContext
import java.io.File

object PrefsUtils {
    const val mPrefsName = "MiuiHome_Config"
    private var mPrefsPathCurrent: String? = null
    private var mPrefsFileCurrent: String? = null
    var mSharedPreferences: SharedPreferences? = null
    private const val mPrefsPath = "/data/user_de/0/" + Helpers.mAppModulePkg + "/shared_prefs"
    private const val mPrefsFile = "$mPrefsPath/$mPrefsName.xml"

    @SuppressLint("WorldReadableFiles")
    fun getSharedPrefs(context: Context, protectedStorage: Boolean, multiProcess: Boolean): SharedPreferences {
        var c = context
        if (protectedStorage) c = getProtectedContext(c)
        return try {
            c.getSharedPreferences(
                mPrefsName, if (multiProcess) Context.MODE_MULTI_PROCESS or Context.MODE_WORLD_READABLE else Context.MODE_WORLD_READABLE
            )
        } catch (t: Throwable) {
            c.getSharedPreferences(mPrefsName, if (multiProcess) Context.MODE_MULTI_PROCESS or Context.MODE_PRIVATE else Context.MODE_PRIVATE)
        }
    }

    fun getSharedPrefs(context: Context, protectedStorage: Boolean): SharedPreferences {
        return getSharedPrefs(context, protectedStorage, false)
    }

    val sharedPrefsPath: String?
        get() = if (mPrefsPathCurrent == null) try {
            val mFile = mSharedPreferences!!.javaClass.getDeclaredField("mFile")
            mFile.isAccessible = true
            mPrefsPathCurrent = (mFile[mSharedPreferences] as File).parentFile?.absolutePath
            mPrefsPathCurrent
        } catch (t: Throwable) {
            print("Test$t")
            mPrefsPath
        } else mPrefsPathCurrent
    val sharedPrefsFile: String?
        get() {
            if (mPrefsFileCurrent == null) return try {
                val fFile = mSharedPreferences!!.javaClass.getDeclaredField("mFile")
                fFile.isAccessible = true
                mPrefsFileCurrent = (fFile[mSharedPreferences] as File).absolutePath
                println("Test: mPrefsFileCurrent")
                mPrefsFileCurrent
            } catch (t: Throwable) {
                println("Test: mPrefsFile$t")
                mPrefsFile
            } else println("Test: mPrefsFileCurrent2")
            return mPrefsFileCurrent
        }
}